File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        i, j = 0, len(height) - 1
        maxHeight = 0

        while i < j:
            maxHeight = max(maxHeight, min(height[i], height[j]) * (j - i))
            if height[i] > height[j]:
                j -= 1
            else:
                i += 1
        
        return maxHeight

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int result = 0;
        int l = 0, r = height.size() - 1;

        while (l < r) {
            int area = (r - l) * std::min(height[l], height[r]);
            result = std::max(result, area);
            if (height[l] < height[r]) {
                l++;
            } else if (height[l] == height[r]) {
                if (height[l + 1] > height[r - 1]) {
                    l++;
                } else {
                    r--;
                }
            } else {
                r--;
            }
        }

        return result;
    }
};